adala/runtimes/_litellm.py

```diff
@@ -0,0 +1,537 @@
+        field_schema = field_schema or {}
+
+        options = {}
+        for field, schema in field_schema.items():
+            if schema.get('type') == 'array':
+                options[field] = schema.get('items', {}).get('enum', [])
+
+        output_fields = parse_template(
+            partial_str_format(output_template, **extra_fields),
+            include_texts=True,
+        )
+        system_prompt = instructions_template
+        user_prompt = input_template.format(**record, **extra_fields)
+        messages = [{'role': 'system', 'content': system_prompt}]
+
+        outputs = {}
+        for output_field in output_fields:
+            if output_field['type'] == 'text':
+                if user_prompt is not None:
+                    user_prompt += f"\n{output_field['text']}"
+                else:
+                    user_prompt = output_field['text']
+            elif output_field['type'] == 'var':
+                name = output_field['text']
+                messages.append({'role': 'user', 'content': user_prompt})
+                completion_text = self.execute(messages)
+                if name in options:
+                    completion_text = match_options(
+                        completion_text, options[name]
+                    )
+                outputs[name] = completion_text
+                messages.append(
+                    {'role': 'assistant', 'content': completion_text}
+                )
+                user_prompt = None
+
+        return outputs
+
+
+class AsyncLiteLLMChatRuntime(AsyncRuntime):
+    """
+    Runtime that uses [OpenAI API](https://openai.com/) and chat completion models to perform the skill.
+    It uses async calls to OpenAI API.
+
+    Attributes:
+        openai_model: OpenAI model name.
+        openai_api_key: OpenAI API key. If not provided, will be taken from OPENAI_API_KEY environment variable.
+        base_url: Can point to any implementation of the OpenAI API. Defaults to OpenAI's.
+        max_tokens: Maximum number of tokens to generate. Defaults to 1000.
+        temperature: Temperature for sampling, between 0 and 1. Higher values means the model will take more risks.
+            Try 0.9 for more creative applications, and 0 (argmax sampling) for ones with a well-defined answer.
+            Defaults to 0.0.
+
+        concurrent_clients: Number of concurrent clients to OpenAI API. More clients means more parallel requests, but
+            also more money spent and more chances to hit the rate limit. Defaults to 10.
+    """
+
+    model_config = ConfigDict(
+        arbitrary_types_allowed=True
+    )  # for @computed_field
+
+    model: str
+    api_key: Optional[str] = None
+    base_url: Optional[str] = None
+    max_tokens: Optional[int] = 1000
+    temperature: Optional[float] = 0.0
+    splitter: Optional[str] = None
+    # concurrent_clients: Optional[int] = 10
+    timeout: Optional[int] = 10
+
+    # @computed_field
+    # def _client(self) -> AsyncOpenAI:
```

Comment:
can we get rid of this?

Reply:
yep ofc, just left it as a reference for now until I commit to getting rid of concurrent clients